src/main/java/graphql/execution/instrumentation/Instrumentation.java

```diff
@@ -0,0 +1,77 @@
+package graphql.execution.instrumentation;
+
+import graphql.ExecutionResult;
+import graphql.execution.ExecutionContext;
+import graphql.language.Document;
+import graphql.schema.DataFetcher;
+import graphql.schema.DataFetchingEnvironment;
+import graphql.schema.GraphQLFieldDefinition;
+import graphql.validation.ValidationError;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Provides the capability to instrument the execution steps of a GraphQL query.
+ *
+ * For example you might want to track which fields are taking the most time to fetch from the backing database
+ * or log what fields are being asked for.
+ */
+public interface Instrumentation {
+
+    /**
+     * This is called just before a query is executed and when this step finishes the {@link InstrumentationContext#onEnd(Object)}
+     * will be called indicating that the step has finished.
+     *
+     * @param requestString the GraphQL query string
+     * @param operationName the GraphQL operation name
+     * @param context the context object in play
+     * @param arguments the arguments in play
+     * @return a non null {@link InstrumentationContext} object that will be called back when the step ends
+     */
+    InstrumentationContext<ExecutionResult> beginExecution(String requestString, String operationName, Object context, Map<String, Object> arguments);
+
+    /**
+     * This is called just before a query is parsed and when this step finishes the {@link InstrumentationContext#onEnd(Object)}
+     * will be called indicating that the step has finished.
+     *
+     * @param requestString the GraphQL query string
+     * @param operationName the GraphQL operation name
+     * @param context the context object in play
+     * @param arguments the arguments in play
+     * @return a non null {@link InstrumentationContext} object that will be called back when the step ends
+     */
+    InstrumentationContext<Document> beginParse(String requestString, String operationName, Object context, Map<String, Object> arguments);
```

Comment:
Having `Document` passed back and String requestString, String operationName, Object context, Map<String, Object> arguments is a contentious design

On the one hand this allows you to SPY on every thing as it happens and that could be great for better instrumentation.

On the other hand it means the parameters are no API and cant easily be changed.

I think we either do NONE or have a Builder of InstrumentationParseParameters so we can add extra parameters in the future.  In fact as I type... this is what I think it should be.

Or.... we have a `Map<String,Object>` arangement which is not at all type safe but infinitely extendable.  Its quasi secret information and might change at any time.

Thoughts?